Fix #170: Use a private gevent threadpool

tiagocoutinho opened the pull request:
Use a private ThreadPool (previous implementation was
patching default gevent threadpool which might prevent
other libraries using the same threadpool from working
properly together with PyTango)

vxgmichel commented on tango/gevent_executor.py:
```diff
@@ -49,18 +50,13 @@ def set_global_executor(executor):
-    raised in the loop to the AsyncResult.
-    """
-    threadpool = gevent.get_hub().threadpool
-    if gevent.version_info < (1, 1) and not hasattr(threadpool, '_spawn'):
```
It seems you got rid of the version checking, is it on purpose?

vxgmichel commented on tango/gevent_executor.py:
```diff
@@ -49,18 +50,13 @@ def set_global_executor(executor):
+    try:
+        return _THREAD_POOL
+    except NameError:
+        _THREAD_POOL = ThreadPool(10)
```
Do we really need this limit of 10?

tiagocoutinho replied:
No, you are right. I will make it unlimited

vxgmichel reviewed (commented):
Thanks!

vxgmichel commented on tango/gevent_executor.py:
```diff
@@ -89,14 +85,14 @@ def get_with_exception(result, block=True, timeout=None):
+        fn = wrap_errors(fn)
+        result = super(ThreadPool, self).spawn(fn, *args, **kwargs)
+        result._get = result.get
+        result.get = types.MethodType(get_with_exception, result, type(result))
```
`type(result)` argument needs to be removed, it's not compatible with python 3 (and it's not mandatory with python 2)

vxgmichel commented on tango/gevent_executor.py:
```diff
@@ -49,18 +50,13 @@ def set_global_executor(executor):
+    global _THREAD_POOL
+    try:
+        return _THREAD_POOL
+    except NameError:
```
Why not use `if _THREAD_POOL is None:` just like in `get_global_executor`?

tiagocoutinho replied:
Ok

tiagocoutinho commented:
FYI, on gevent/gevent#55 they propose to configure this behaviour at the application level. 

If, for example, you want to suppress greenlet exception printing you can configure what gevent considers to be an error by doing something like this:
```python
import gevent.hub
gevent.hub.Hub.NOT_ERROR=(Exception,)
```
For, PyTango, the fact that we ask the tango method to be executed in a gevent threadpool is an implementation detail so the approach to wrap tango call errors is the correct one.

Another info is that the code to wrap exceptions exists in gevent itself `gevent.util.wrap_errors` but both the location and implementation have changed in recent versions so we prefer to keep the wrap logic inside PyTango rather then rely on an unstable API.

tiagocoutinho committed:
Fix #170: prevent gevent from calling sys.excepthook

Use a private ThreadPool (previous implementation was
patching default gevent threadpool which might prevent
other libraries using the same threadpool from working
properly together with PyTango)

vxgmichel committed:
Maxsize is a mandatory argument of gevent ThreadPool

vxgmichel committed:
Fix gevent 1.3 compatibility

vxgmichel commented:
@tiagocoutinho Thanks!
